### ISSUE TITLE
Rename Required Actions (Install, Uninstall)

### ISSUE THREAD

**Issue by astrieanna** [cnabio/cnab-spec#204]
I think that Install and Uninstall are not the best terms for the required actions and that they are a poor enough fit that they cause problems communicating with stakeholders (creators of CNAB bundles, etc). If a change can be agreed to, the time to make the change is now (pre-1.0).

Install/Uninstall makes some implications:
1. When you install software, it is to a specific laptop or server (e.g. installing Chrome).
2. You can install something once, upgrade/etc many times, then uninstall once.
3. An "installation" is a physical art-piece or physical infrastructure; it is not generally used to describe software.

The first point lead to confusion for me when I was attempting to understand what CNAB is for and what the runtime does with the invocation image. I expected the invocation image to be run on a specific server/container to prepare it to run some other software (the way kubeadm runs to turn a server into a kubernetes node).

The second point leads to confusion for potential CNAB creators about what use cases a CNAB bundle can fill. A more generic term would help people think more expansively about how to use CNAB.

Installation is also a confusing term. It is less urgent to change this because it is not a term embedded in the implementations the way install and uninstall are (as far as I know), but I wanted to mention it because a replacement term is likely to be influenced by the replacement for install/uninstall.

Some examples of generic verb pairs that I would be happy with:
* create/delete
* create/destroy
* provision/deprovision
* deploy/destroy
* start/end
* launch/end

----
Addition words (expanding scope):
* update or upgrade for third verb
* what to call the "thing a bundle installs/creates/etc" -- currently installation

cc @ryanmoran

**Comment by simonferquel** [comment]
I'd rule out create/delete as it is confusing if you want to create a bundle, or create an installation.

Actually my preference is still install/uninstall with a short second place for provision/deprovision

**Comment by glyn** [comment]
> I'd rule out create/delete as it is confusing if you want to create a bundle, or create an installation.
> 
> Actually my preference is still install/uninstall with a short second place for provision/deprovision

+1

Also, we should consider what to do with "upgrade" as I think this may carry install/uninstall connotations for some people. For example, is provision/upgrade/deprovision" reasonable? I think it's a bit clunky.

**Comment by maybo** [comment]
Hi, I'm usually read-only, but we are in very similar discussions :)
we decided to go for  "Start"/"launch" and "End" for our bundle (btw we are looking at how to work with CNAB). The reason was that "provisioning" is saved for infrastrcture-related defintions, "deployment" and "install" usually refer to applications and we also got feedback that it would be extremely confusing and limiting to use these terminologies for a bundle that represents both. Another reason that we went for this terminology was that we put strong emphais on the lifecycle of the bundle (we think is very importat, I don't know what your view is on this right now)

**Comment by astrieanna** [comment]
🥇 I really like those word pairs because they match how generic bundles are and focus on when the phases run. I've added them to the list at the top to consolidate the options.

**Comment by glyn** [comment]
@astrieanna What would you propose to do with "upgrade" for each of those pairs?

**Comment by maybo** [comment]
@glyn we went for Launch/End/Update

**Comment by simongdavies** [comment]
If this is going to change then I would go for create\upgrade\destroy. It conveys that an instance of the application and any associated resources\services\data is being created, upgraded and destroyed. 

I don't think that there is an issue with confusion with creating a bundle, the use of create is defined by the tool not by the spec and could be changed to scaffold (the help description for create in duffle already uses scaffold not create i.e. Create - Scaffold a bundle ), alternatively new could be used by tools.

**Comment by glyn** [comment]
I'd like to understand how we will talk about the actions informally and in documentation. I'd be much more likely to talk about installing a CNAB rather than creating a CNAB since that really could be confused with, er, creating the CNAB.

**Comment by astrieanna** [comment]
@glyn I think the confusion you're pointing is that if you "create a CNAB", you should end up with a new CNAB. If you "run the create action on a CNAB" (too many words), you should end up with a <thing this bundle creates>.

I feel like the verbs for these action names are the things you're asking the invocation image to do, so you don't need the "<verb> a CNAB" form to work -- you need the "<verb> a <thing a CNAB creates>" to work.

In general, I think using the "install/upgrade/uninstall a CNAB" is confusing because there's two things to manage: the bundle itself (i.e. an actual thick bundle) and the things created by the bundle. It can already be confusing what it means to "uninstall a CNAB" -- are you managing the CNAB itself or something the CNAB made?

**Comment by ryanmoran** [comment]
I think that I find the phrasing of installing a bundle to be a bit confusing. "Installation" is the word that the spec uses to describe the instance of a "running" bundle. But the only open runtime that I know of, `duffle` makes no mention of that concept at all. Instead, it opts for using a command like `duffle bundle install` to handle this process.

I tend to think of it as "installing a bundle" == "creating an installation".

**Comment by glyn** [comment]
Those are good points @astrieanna and @ryanmoran. I'm wondering what informal language we'll use, regardless of the action names. Some options are:
* run the X action on a CNAB (too wordy)
* run X on a CNAB
* X a CNAB (this seems to be the most confusing and yet the one we'll fall into unless we are careful)

On a separate point, if we do rename the standard actions perhaps we should add a standard prefix to them, e.g. cnab_provision, cnab_update, cnab_deprovision and prohibit custom actions from using that prefix. This would allow us to add more standard actions later without risking overlapping custom actions in existing bundles.

**Comment by technosophos** [comment]
Just to clarify some roots: Terminology is largely around the root "instance". The cognates "install" and "installation" are related by design. "If you install a bundle, you get an instance of that bundle". The term "install" NEVER refers to bundle as a _class_, but as an _instance_. (I don't _install_ a CNAB  bundle into a registry, I _push_ it into a registry. When I am done, I _delete_ it from a registry).

- Instance: A specific concrete realization of a bundle (r.e. in class theory, the bundle is the class, and the instance is the instance.)
- Installation: Is largely used as a synonym for instance
- Install: To create an instance (or to _instantiate_, if you like that term better)
- Uninstall: Reverse the process of installing

Initially, we used `Release` instead of `Installation` (following another devops pattern of speaking), but it was misleading. So if there are any vestigial uses of that term, we should replace it with `instance`.

Terms like `create`, `apply`, `provision`.... none of those really speak innately to the difference between class/instance distinction, which is what the CNAB and bundle instance concepts are supposed to capture.

**Comment by technosophos** [comment]
I am closing this issue since CNAB Core 1.0 shipped. We cannot change it at this point.
